---
title: "QTL Identification and Analysis with QTLretrievR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{QTL Identification and Analysis with QTLretrievR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<div align='center'>**Abstract**</div>
Quantitative trait locus (QTL) analysis is a frequently used technique in omics studies to identify molecular phenotypes that have effects on what is being measured. These analyses are often computationally intensive and intimidating to those without a strong computational background, especially when samples are taken for multiple cell types that wonâ€™t be analyzed concurrently. QTLretirevR is designed to simplify and streamline the process of identifying QTL peaks, mediators of those peaks, and specific local effects. Built on `qtl2` (for peak calling and effect identification) and `intermediate` (for peak mediation), QTLretrievR is designed to use nested parallelism in order to optimize the amount of time it takes to run each analysis step. In its first version it is designed to work with raw genotype probabilities fromgenome reconstruction by RNA-Seq (GBRS) (or any qtl2 formatted genome probabilities) for all samples, and normalized quantification data for Collaborative Cross and Diversity Outbred mice.

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(QTLretrievR)
```

# Standard workflow

## Quick start

This is the most basic way to run QTLretrievR. In order to use QTLretievR there are some specific files that need to be generated in preparation for identifying QTL peaks and the further analysis. Genotyping files are necessary for each of the samples and normalized and filtered quantification data for each cell type or confounding factor collated into a single file. The quantified data can be passed either as a path to a file, or an R object. Other information that is required is sample metadata, which variable within the sample metadata are covariate factors that need to be accounted for when peak calling, and an annotations that contains information about the genes or proteins that are being measured.

Please note in the example below that the objects will be only be returned, not saved. If you wish to have the objects saved for later use you can change `save` to "so" (save only), or "sr" (save and return).

```{r quick_start}
## Passing a Genoprobs object - list of probabilities for each tissue
## Using internal data for expr, gridFile, metadata, biomart

qtl_res <- runQTL(outdir = "../../vignette/", gbrs_fileLoc = list("mESC" = demo_probs),
                  geno_out = "mESC_mm10_probs.rds", peaks_out = "mESC_mm10_peaks.rds", 
                  map_out = "mESC_mm10_map.rds", med_out = "mESC_mm10_mediation.rds",
                  effects_out = "mESC_mm10_effects.rds",
                  expr_mats = list("mESC" = demo_counts),
                  tissues = "mESC", gridFile = gridfile69k, covar_factors = c("sex"),
                  metadata = demo_meta, biomart = demo_annot, save_t = "ro")
```

This can be alternatively submitted as follows:
```{r quick_start_file_in, eval = FALSE}
## Unevaluated Code Chunk

qtl_res <- runQTL(outdir = "../../vignette/", gbrs_fileLoc = <path/to/gbrs/files>,
                  geno_out = "mESC_mm10_probs.rds", peaks_out = "mESC_mm10_peaks.rds", 
                  map_out = "mESC_mm10_map.rds", med_out = "mESC_mm10_mediation.rds",
                  effects_out = "mESC_mm10_effects.rds", expr_mats = c(<path/to/expr_mat1>, <path/to/expr_mat2>),
                  tissues = "mESC", gridFile = <path/to/gridfile>, covar_factors = c("sex"),
                  metadata = <path/to/sample/metadata>, biomart = <path/to/annotations>, save = "ro"
                  )

```
This returns the mapping object, original peaks object, mediation object, and effects object. The peaks, effects and mediation objects can be used for additional plotting and analysis.

## Input Data

There are several types of data that `QTLretrievR` takes as input, below we describe variations on the required and optional data inputs.

### Genotype Probabilities

Genotype Probabilities (genoprobs) can be passed to `QTLretreivR` in a few different ways. 

  - Passed as a string to the location .tsv probability files from the GBRS pipeline. This can be the folder where all the GBRS results are kept, the files should end with `.gbrs.interpolated.genoprobs.tsv`. This option would be used with `genoprobably` and `runQTL`. This is probably the easiest option if you are starting from scratch.
  - Passed as a two dimensional array of the genotype probabilities for all the samples. See the `genotype_probs.Rds` file [here](https://figshare.com/articles/dataset/Supplemental_data_repository_for_Skelly_et_al_2020_Cell_Stem_Cell_Mapping_the_effects_of_genetic_variation_on_chromatin_state_and_gene_expression_reveals_loci_that_control_ground_state_pluripotency_/12233570) for an example of how this should look. This option is for passing to `runQTL`.
  - Passed as a `qtl2` formatted genoprobs object. This option is for passing to `runQTL` and `mapQTL`. An example of this is shown below.
  
```{r probs_ex}
## Structure and example of qtl2 fomatted genoprobs
str(demo_probs)
demo_probs[[1]][1:5,,1:5]
```
  
### Quantified Data

Although the variable is named `expr_mats` any quantified data can be passed to this variable. They can be passed as a vector of paths to the data (in the same order as tissues are being passed), or as a named list of matrices (named with the tissue type associated with the counts). These should be normalized data with the phenotype as row names and the samples as columns. See the attached `demo_counts` for an example of formatting.

```{r counts_ex}
## Information and partial example of quantified data
class(demo_counts)
head(demo_counts)
```

### Marker Grid

This should be a grid of markers and the associated genomic locations in bp and cM. The grid file can be passed as an object, or as a path to where the file is located. There are two grid files attached to `QTLretrievR`, a grid with ~75K markers with positions updated for mm38 and a grid with ~69K markers with positions associated with mm10. If your probabilities are derived from a MUGA scan, the markers can be interpolated for the 75K grid, or you can make a marker grid to be used with your probabilities. If making your own grid file, see `gridfile` to make sure you have the correct format

```{r marker_ex}
## example format for 75K marker grid
head(gridfile)
```

### Metadata

At minimum the metadata file should contain a column for sample IDs (`ID`) and columns for any covariates to be included in the peak and effect calling. Any other pieces of information that may be relevant to your analysis that you want to make sure is all in the same place should also be included. A minimal example, `demo_meta`, is attached to `QTLretrievR`.

```{r meta_ex}
## minimal metadata example
head(demo_meta[which(demo_meta$ID %in% colnames(demo_counts)),])
```

### Annotations

Currently the annotations needs to contain at minimum a `gene`, `symbol`, `chr`, `start`, and `end` column. The `gene` column should contain the phenotype information that appears in the row names of the quantified data. In the attached example (`demo_annot`) the gene column contains ensembl gene IDs, these could be traded out for ensembl protein IDs or any other phenotype identifier.

```{r annot_ex}
## Example annotation file - Note this file is all the genes for this biomart build
head(demo_annot)
```

## Running as Separate Steps

If, instead of using the wrapper to run all steps at once, you wish to run each step individually (calculating/formatting the genotype probabilities, mapping/peak calling, mediation, effects) this can be done with relative ease.

### Genoprobs

In order to convert the genoprobs from the GBRS output to the `qtl2` format we run a function called `genoprobably`. It pulls in the relevant files from the provided location and splits samples into the tissues they are derived from (the names that you pass to `tissue` should be included in the file name in some form), then puts these into a 3 dimensional format based on the markers. Once the genotype probabilities are in the 3D format `probs_3d_to_qtl2` is used to convert them to the qtl2 format. The below example shows how to run genoprobably with the 75k gridfile.

```{r genoprobably, eval=FALSE}
## Unevaluated Code Chunk
probs <- genoprobably(outfile = "../../vignette/gbrs_interpolated_probs.rds", 
                      gbrsFileLoc = "<path/to/gbrs/results>",
                      tissues = c("tissue1","tissue2"), gridFile = gridfile, save = "ro")
```

### Mapping and Peak Calling

The mapping and peak calling steps are combined in our pipeline. We use the genoprobs to calculate kinship, and create the genetic and pysical maps. Then using the quantified data and sample metadata we rankZ normalize the counts, calculate covariates and finally do the peak calling using `qtl2::scan1`.

```{r mapQTL, eval=FALSE}
## Unevaluated Code Chunk
## Passing objects:
map_peaks <- mapQTL(outdir = "../../vignette/", peaks_out = "mm39_peaks.rds",
                    map_out = "mm39_map.rds", genoprobs = probs, samp_meta = demo_meta,
                    expr_mats = list("mESC" = demo_counts), covar_factors = c("sex"),
                    gridFile = gridfile, biomart = demo_annot, save = "ro")

## Passing file locations
map_peaks <- mapQTL(outdir = "../../vignette/", peaks_out = "mm39_peaks.rds",
                    map_out = "mm39_map.rds", genoprobs = "path/to/saved/probs.rds",
                    samp_meta = "path/to/sample/metadata",
                    expr_mats = c("path/to/counts1","path/to/coounts/2"), 
                    covar_factors = c("sex"),
                    gridFile = "path/to/custom/grid",
                    biomart = "path/to/annotations", save = "ro")
```

### Mediation

For mediation we use the peaks, mapping information (which includes the rankZ normalized counts), and annotations to identify targets which in turn allows us to identify phenotypes that act as mediators for peaks above a suggestive LOD. This is done using `intermediate::mediation.scan`.

```{r qtl_mediate, eval=FALSE}
## Unevaluated Code Chunk
med_res <- run_mediate(peaks = map_peaks$peaks_list, mapping = map_peaks$maps_list,
                       outdir = "../../vignette", biomart = demo_annot,
                       suggLOD - 7, med_out = "mm39_mediation.rds", save = "ro")
```

### Effects

For mediation we use the peaks, mapping information (which includes the rankZ normalized counts), and annotations to identify targets which in turn allows us to identify phenotypes that act as mediators for peaks above a suggestive LOD. This is done using `intermediate::mediation.scan`.

```{r qtl_effects, eval=FALSE}
## Unevaluated Code Chunk
effects <- qtl_effects(mapping = map_peaks$maps_list, peaks = map_peaks$peaks_list,
                       suggLOD = 8, outdir = "../../vignette", 
                       outfile = "mm39_effects.rds", save = "ro")
```
