<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>QTL Identification and Analysis with QTLretrievR • QTLretrievR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="QTL Identification and Analysis with QTLretrievR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">QTLretrievR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/QTLretrievR.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/DOAttie500_LiverLipids.html">Liver Lipids Example</a></li>
    <li><a class="dropdown-item" href="../articles/DOAttie500_Islet.html">Islet Transcriptomics Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>QTL Identification and Analysis with QTLretrievR</h1>
            
      

      <div class="d-none name"><code>QTLretrievR.Rmd</code></div>
    </div>

    
    
<div align="center">
<strong>Abstract</strong>
</div>
<p>Advances in molecular phenotyping technologies, including
quantitative transcriptomics and proteomics (“-omics”), have transformed
biomedical research over the past two decades by providing a granular
understanding of the gene regulatory networks that drive disease. Many
studies have demonstrated the additional power of combining -omics
profiling with genetic mapping, an integrative approach referred to as
“systems genetics”, to link population variability in complex disease
phenotypes back to genetic variants and their direct effects on gene
regulation (e.g. gene expression). However, these quantitative trait
locus (QTL) analyses are often computationally intensive and
intimidating to those without a strong computational background —
especially when samples originate from multiple cell/tissue types or
involve thousands of molecular phenotypes. We have developed a new R
package, QTLretrievR, that aims to simplify and streamline the
identification and downstream analyses of molecular QTL from genetically
diverse populations such as Diversity Outbred or Collaborative Cross
mice. QTLretrievR leverages existing R packages used in complex trait
analysis, including the popular <code>qtl2</code> package for QTL peak
calling and haplotype effect identification and the
<code>intermediate</code> package for peak mediation, and employs
dynamic nested parallelism to enhance the efficiency of each analysis
step. Future developments will enhance the visualization functionality
and data analysis options, including adding network inference methods
and expanding the software to work with other mapping populations.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://deweyhannah.github.io/QTLretrievR/">QTLretrievR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>This is the easiest way to run QTLretrievR. There are several input
files required for <code>runQTL</code> as described in detail below.
This function will return the list of all the objects used in QTL
mapping, annotated QTL peaks, QTL effects, and mediation results. Note
in the example below that the objects will be only be returned, not
saved. If you wish to have the objects saved for later use you can
change <code>save</code> to “so” (save only), or “sr” (save and
return).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Passing a Genoprobs object - list of probabilities for each tissue</span></span>
<span><span class="co">## Using internal data for expr, gridFile, metadata, annots</span></span>
<span><span class="co">## Unevaluated Code Chunk</span></span>
<span></span>
<span><span class="va">qtl_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runQTL.html">runQTL</a></span><span class="op">(</span>outdir <span class="op">=</span> <span class="st">"../../vignette/"</span>, </span>
<span>                  gbrs_fileLoc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"mESC"</span> <span class="op">=</span> <span class="va">demo_probs</span><span class="op">)</span>,</span>
<span>                  geno_out <span class="op">=</span> <span class="st">"mESC_mm10_probs.rds"</span>, </span>
<span>                  peaks_out <span class="op">=</span> <span class="st">"mESC_mm10_peaks.rds"</span>, </span>
<span>                  map_out <span class="op">=</span> <span class="st">"mESC_mm10_map.rds"</span>, </span>
<span>                  med_out <span class="op">=</span> <span class="st">"mESC_mm10_mediation.rds"</span>,</span>
<span>                  effects_out <span class="op">=</span> <span class="st">"mESC_mm10_effects.rds"</span>,</span>
<span>                  expr_mats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"mESC"</span> <span class="op">=</span> <span class="va">demo_counts</span><span class="op">)</span>,</span>
<span>                  tissues <span class="op">=</span> <span class="st">"mESC"</span>, </span>
<span>                  gridFile <span class="op">=</span> <span class="va">gridfile69k</span>, </span>
<span>                  covar_factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span><span class="op">)</span>,</span>
<span>                  metadata <span class="op">=</span> <span class="va">demo_meta</span>, </span>
<span>                  annots <span class="op">=</span> <span class="va">demo_annot</span>, </span>
<span>                  save_t <span class="op">=</span> <span class="st">"ro"</span><span class="op">)</span></span></code></pre></div>
<p>This can be alternatively submitted as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">## Unevaluated Code Chunk</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>qtl_res <span class="ot">&lt;-</span> <span class="fu">runQTL</span>(<span class="at">outdir =</span> <span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>output<span class="sc">/</span>dir<span class="sc">&gt;</span>, </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                  <span class="at">gbrs_fileLoc =</span> <span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>gbrs<span class="sc">/</span>files<span class="sc">&gt;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>                  <span class="at">geno_out =</span> <span class="st">"Name_for_genotype_prob_file.rds"</span>, </span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>                  <span class="at">peaks_out =</span> <span class="st">"Name_for_QTL_peaks.rds"</span>, </span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>                  <span class="at">map_out =</span> <span class="st">"Name_for_mapping_objects.rds"</span>, </span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>                  <span class="at">med_out =</span> <span class="st">"Name_for_mediation_Results.rds"</span>,</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>                  <span class="at">effects_out =</span> <span class="st">"Name_for_QTL_effects.rds"</span>,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>                  <span class="at">expr_mats =</span> <span class="fu">c</span>(<span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>expr_mat1<span class="sc">&gt;</span>, <span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>expr_mat2<span class="sc">&gt;</span>),</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>                  <span class="at">tissues =</span> <span class="st">"Tissue_name"</span>,</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>                  <span class="at">gridFile =</span> <span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>gridfile<span class="sc">&gt;</span>, </span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>                  <span class="at">covar_factors =</span> <span class="fu">c</span>(<span class="st">"Covariates"</span>),</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>                  <span class="at">metadata =</span> <span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>sample<span class="sc">/</span>metadata<span class="sc">&gt;</span>, </span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>                  <span class="at">annots =</span> <span class="sc">&lt;</span>path<span class="sc">/</span>to<span class="sc">/</span>annotations<span class="sc">&gt;</span>, </span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>                  <span class="at">save =</span> <span class="st">"ro"</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>                  )</span></code></pre></div>
<p>This returns the mapping object, original peaks object, mediation
object, and effects object. The peaks, effects and mediation objects can
be used for additional plotting and analysis.</p>
</div>
<div class="section level2">
<h2 id="inputs-data">Inputs Data<a class="anchor" aria-label="anchor" href="#inputs-data"></a>
</h2>
<p>There are several types of data that <code>QTLretrievR</code> takes
as input, below we describe variations on the required and optional data
inputs.</p>
<div class="section level3">
<h3 id="genotype-probabilities">Genotype Probabilities<a class="anchor" aria-label="anchor" href="#genotype-probabilities"></a>
</h3>
<p>Genotype Probabilities (genoprobs) can be passed to
<code>QTLretreivR</code> in a few different ways.</p>
<ul>
<li><p>Passed as a string to the location of the .tsv probability files
from the GBRS software. This can be the folder where all the GBRS
results are kept, the files should end with
<code>.gbrs.interpolated.genoprobs.tsv</code>. This option would be used
with <code>genoprobably</code> and <code>runQTL</code>. This is probably
the easiest option if you are starting from scratch and only have GBRS
genotypes.</p></li>
<li><p>Passed as a two dimensional array of the genotype probabilities
for all the samples. See the <code>genotype_probs.Rds</code> file <a href="https://figshare.com/articles/dataset/Supplemental_data_repository_for_Skelly_et_al_2020_Cell_Stem_Cell_Mapping_the_effects_of_genetic_variation_on_chromatin_state_and_gene_expression_reveals_loci_that_control_ground_state_pluripotency_/12233570" class="external-link">here</a>
for an example of how this should look. This option is for passing to
<code>runQTL</code>.</p></li>
<li><p>Passed as a <code>qtl2</code> formatted genoprobs object. This
option is for passing to <code>runQTL</code> and <code>mapQTL</code>. An
example of this is shown below.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Structure and example of qtl2 fomatted genoprobs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">demo_probs</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14" "15"</span></span>
<span><span class="co">#&gt; [16] "16" "17" "18" "19" "X"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">demo_probs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;  num [1:10, 1:8, 1:4711] 1.48e-08 1.23e-08 1.04e-08 3.31e-08 2.94e-07 ...</span></span>
<span><span class="co">#&gt;  - attr(*, "dimnames")=List of 3</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:10] "PB357.02_repA" "PB357.12_repA" "PB357.16_repA" "PB357.18_repA" ...</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:8] "A" "B" "C" "D" ...</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:4711] "1_3000000" "1_3041392" "1_3346528" "1_3651663" ...</span></span>
<span><span class="va">demo_probs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="co">#&gt; , , 1_3000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                          A   B            C            D            E</span></span>
<span><span class="co">#&gt; PB357.02_repA 1.477899e-08 0.5 1.217364e-08 1.370868e-08 1.199720e-08</span></span>
<span><span class="co">#&gt; PB357.12_repA 1.232415e-08 0.5 5.000000e-01 1.604246e-08 3.499095e-09</span></span>
<span><span class="co">#&gt;                          F            G            H</span></span>
<span><span class="co">#&gt; PB357.02_repA 5.000000e-01 4.331041e-09 1.313560e-08</span></span>
<span><span class="co">#&gt; PB357.12_repA 1.478095e-08 4.742920e-09 1.118048e-08</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; , , 1_3041392</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                          A   B            C            D            E</span></span>
<span><span class="co">#&gt; PB357.02_repA 1.477899e-08 0.5 1.217364e-08 1.370868e-08 1.199720e-08</span></span>
<span><span class="co">#&gt; PB357.12_repA 1.232415e-08 0.5 5.000000e-01 1.604246e-08 3.499095e-09</span></span>
<span><span class="co">#&gt;                          F            G            H</span></span>
<span><span class="co">#&gt; PB357.02_repA 5.000000e-01 4.331041e-09 1.313560e-08</span></span>
<span><span class="co">#&gt; PB357.12_repA 1.478095e-08 4.742920e-09 1.118048e-08</span></span></code></pre></div>
<p>If you have genotyping from one of the Mouse Universal Genotyping
Arrays (MUGA), then you can format your probabilities using our
<code>mugaprobs</code> function, a detailed example is below in the <a href="#genoprobs">Genoprobs</a> section of this document.</p>
</div>
<div class="section level3">
<h3 id="quantified-phenotypes">Quantified Phenotypes<a class="anchor" aria-label="anchor" href="#quantified-phenotypes"></a>
</h3>
<p>Although the variable is named <code>expr_mats</code> any quantified
molecular data can be passed to this variable. They can be passed as a
vector of paths to the data (in the same order as tissues are being
passed), or as a named list of matrices (named with the tissue type
associated with the counts). These should be filtered, batch corrected
and normalized data with the phenotype in rows and the samples in
columns, there is no need to rankZ transform your counts, that step
occurs in <code>mapQTL</code>. See the attached <code>demo_counts</code>
for an example of formatting.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Information and partial example of quantified data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">demo_counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "matrix" "array"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demo_counts</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    PB357.02_repA PB357.12_repA PB357.16_repA PB357.18_repA</span></span>
<span><span class="co">#&gt; ENSMUSG00000030402      90.27620      47.49881      65.92215      60.68472</span></span>
<span><span class="co">#&gt; ENSMUSG00000025473      55.33058      61.24847      59.68627      26.61611</span></span>
<span><span class="co">#&gt; ENSMUSG00000029246    2333.74400    5135.62454    2341.10901    2701.05069</span></span>
<span><span class="co">#&gt; ENSMUSG00000051413     218.41017     162.49593     209.34736     223.57530</span></span>
<span><span class="co">#&gt; ENSMUSG00000026356    3244.11907   11302.21694    2892.56671    4801.55973</span></span>
<span><span class="co">#&gt; ENSMUSG00000042505     317.42278     554.98610     316.24815     300.22969</span></span>
<span><span class="co">#&gt;                    PB357.19_repA PB357.20_repA PB357.21_repA PB357.22_repA</span></span>
<span><span class="co">#&gt; ENSMUSG00000030402       50.6948      43.98085      68.98128      57.27735</span></span>
<span><span class="co">#&gt; ENSMUSG00000025473       24.3725      45.02802      55.41119      37.55891</span></span>
<span><span class="co">#&gt; ENSMUSG00000029246     3481.3853    2611.54250    1373.65252    3451.91477</span></span>
<span><span class="co">#&gt; ENSMUSG00000051413      230.0764     158.12164     211.46720     217.84171</span></span>
<span><span class="co">#&gt; ENSMUSG00000026356     5616.3992    4653.59323    2312.56906    5754.02576</span></span>
<span><span class="co">#&gt; ENSMUSG00000042505      348.0393     347.65818     360.73816     294.83748</span></span>
<span><span class="co">#&gt;                    PB357.28_repA PB357.29_repA</span></span>
<span><span class="co">#&gt; ENSMUSG00000030402      42.11575      52.36949</span></span>
<span><span class="co">#&gt; ENSMUSG00000025473      14.03858      44.38093</span></span>
<span><span class="co">#&gt; ENSMUSG00000029246    3130.74325    2534.28661</span></span>
<span><span class="co">#&gt; ENSMUSG00000051413     161.91166     260.95984</span></span>
<span><span class="co">#&gt; ENSMUSG00000026356    6941.61141    4452.29448</span></span>
<span><span class="co">#&gt; ENSMUSG00000042505     440.81151     286.70078</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="marker-grid">Marker Grid<a class="anchor" aria-label="anchor" href="#marker-grid"></a>
</h3>
<p>This should be a grid of markers and the associated genomic locations
in bp and cM. The grid file can be passed as an object, or as a path to
where the file is located. There are two grid files attached to
<code>QTLretrievR</code>, a grid with ~75K pseudo-markers with positions
updated for mm38 and a grid with ~69K pseudo-markers with positions
associated with mm10, and the GigaMUGA and MegaMUGA markers and
positions. Or you can use a new marker grid to be used with your
probabilities. If making your own grid file, see <code>gridfile</code>
to make sure you have the correct format.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example format for 75K marker grid</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gridfile</span><span class="op">)</span></span>
<span><span class="co">#&gt;           chr     pos      cM</span></span>
<span><span class="co">#&gt; 1_3000000   1 3000000 0.00001</span></span>
<span><span class="co">#&gt; 1_3039563   1 3039563 0.02001</span></span>
<span><span class="co">#&gt; 1_3079126   1 3079126 0.04001</span></span>
<span><span class="co">#&gt; 1_3118689   1 3118689 0.06001</span></span>
<span><span class="co">#&gt; 1_3158252   1 3158252 0.08001</span></span>
<span><span class="co">#&gt; 1_3197814   1 3197814 0.10001</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>At a minimum, the metadata file should contain a column for sample
IDs (<code>ID</code>) and columns for any covariates to be included in
the peak and effect calling. Any other pieces of information that may be
relevant to your analysis that you want to make sure is all in the same
place should also be included. A minimal example,
<code>demo_meta</code>, is attached to <code>QTLretrievR</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## minimal metadata example</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demo_meta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">demo_meta</span><span class="op">$</span><span class="va">ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">demo_counts</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;              ID sex</span></span>
<span><span class="co">#&gt; 1 PB357.02_repA   F</span></span>
<span><span class="co">#&gt; 2 PB357.12_repA   F</span></span>
<span><span class="co">#&gt; 3 PB357.16_repA   F</span></span>
<span><span class="co">#&gt; 4 PB357.18_repA   F</span></span>
<span><span class="co">#&gt; 5 PB357.19_repA   F</span></span>
<span><span class="co">#&gt; 6 PB357.20_repA   F</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotations">Annotations<a class="anchor" aria-label="anchor" href="#annotations"></a>
</h3>
<p>Currently the annotations needs to contain at a minimum
<code>id</code>, <code>symbol</code>, <code>chr</code>,
<code>start</code>, and <code>end</code> columns. The <code>id</code>
column should contain the phenotype information that appears in the row
names of the quantified data. In the attached example
(<code>demo_annot</code>) the id column contains ensembl gene IDs, these
could be traded out for ensembl protein IDs or any other phenotype
identifier. The <code>demo_annot</code> file is the Ensemblv84 biomart
annotations, if you are using newer expression data use
<code>annot_105</code> or upload your own annotations</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example annotation file - Note this file is all the genes for the Ensemblv84 build</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demo_annot</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   id symbol chr     start       end</span></span>
<span><span class="co">#&gt; 1 ENSMUSG00000000001  Gnai3   3 108107280 108146146</span></span>
<span><span class="co">#&gt; 2 ENSMUSG00000000028  Cdc45  16  18780447  18811987</span></span>
<span><span class="co">#&gt; 3 ENSMUSG00000000031    H19   7 142575529 142578143</span></span>
<span><span class="co">#&gt; 4 ENSMUSG00000000037  Scml2   X 161117193 161258213</span></span>
<span><span class="co">#&gt; 5 ENSMUSG00000000056   Narf  11 121237253 121255856</span></span>
<span><span class="co">#&gt; 6 ENSMUSG00000000058   Cav2   6  17281185  17289115</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-as-separate-steps">Running as Separate Steps<a class="anchor" aria-label="anchor" href="#running-as-separate-steps"></a>
</h2>
<p>If, instead of using the wrapper to run all steps at once, you wish
to run each step individually (calculating/formatting the genotype
probabilities, mapping/peak calling, mediation, effects) this can be
done with relative ease.</p>
<div class="section level3">
<h3 id="genoprobs">Genoprobs<a class="anchor" aria-label="anchor" href="#genoprobs"></a>
</h3>
<p>In order to convert the genoprobs from the GBRS output to the
<code>qtl2</code> format we run a function called
<code>genoprobably</code>. It pulls in the relevant files from the
provided location and splits samples into the tissues they are derived
from (the names that you pass to <code>tissue</code> should be included
in the file name in some form), then puts these into a 3 dimensional
format based on the markers. Once the genotype probabilities are in the
3D format <code>probs_3d_to_qtl2</code> is used to convert them to the
qtl2 format. The below example shows how to run genoprobably with the
75k gridfile.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Unevaluated Code Chunk</span></span>
<span><span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genoprobably.html">genoprobably</a></span><span class="op">(</span>outfile     <span class="op">=</span> <span class="st">"../../vignette/gbrs_interpolated_probs.rds"</span>, </span>
<span>                      gbrsFileLoc <span class="op">=</span> <span class="st">"&lt;path/to/gbrs/results&gt;"</span>,</span>
<span>                      tissues     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tissue1"</span>,<span class="st">"tissue2"</span><span class="op">)</span>, </span>
<span>                      gridFile    <span class="op">=</span> <span class="va">gridfile</span>, </span>
<span>                      save        <span class="op">=</span> <span class="st">"ro"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively you can use genotyping from either GigaMUGA or MegaMUGA
as your genome probabilities. In order to convert these reports to the
<code>qtl2</code> format we run a function called
<code>mugaprobs</code>. It loads the appropriate grid files, builds the
control file, and processes the genotyping file(s) into the correct
allele based format for processing. The below example shows how to run
<code>mugaprobs</code> from a sample GigaMUGA genotyping run.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Unevaluated Code Chunk</span></span>
<span></span>
<span><span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mugaprobs.html">mugaprobs</a></span><span class="op">(</span>type       <span class="op">=</span> <span class="st">"GM"</span>,                                   <span class="co"># Did your genotyping come from GigaMUGA or MegaMUGA?</span></span>
<span>                   covarLoc   <span class="op">=</span> <span class="st">"/path/to/covariate/file/"</span>, </span>
<span>                   covar_file <span class="op">=</span> <span class="st">"name_of_covariate_file.tsv"</span>, </span>
<span>                   i.files    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"final_report_1"</span>, <span class="st">"final_report_2"</span><span class="op">)</span>,  <span class="co"># If you already have chromosome specific genotype files you can pass that directory as a string here</span></span>
<span>                   genoPrefix <span class="op">=</span> <span class="st">"gm4qtl2"</span>, </span>
<span>                   probsOut   <span class="op">=</span> <span class="st">"muga_interpolated_genoprobs.rds"</span>,      <span class="co"># This is the file that your probabilites will be saved to</span></span>
<span>                   tissues    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tissue1"</span>, <span class="st">"tissue2"</span><span class="op">)</span><span class="op">)</span>                <span class="co"># All probabilities will be attributed to each tissue, this is just to put it in the format needed for `mapQTL`</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="mapping-and-peak-calling">Mapping and Peak Calling<a class="anchor" aria-label="anchor" href="#mapping-and-peak-calling"></a>
</h3>
<p>The mapping and peak calling steps are combined in our pipeline. We
use the genoprobs to calculate kinship, and create the genetic and
pysical maps. Then using the quantified data and sample metadata we
rankZ normalize the counts, calculate covariates and finally do the peak
calling using <code><a href="https://rdrr.io/pkg/qtl2/man/scan1.html" class="external-link">qtl2::scan1</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Unevaluated Code Chunk</span></span>
<span><span class="co">## Passing objects:</span></span>
<span><span class="va">map_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapQTL.html">mapQTL</a></span><span class="op">(</span>outdir        <span class="op">=</span> <span class="st">"../../vignette/"</span>, </span>
<span>                    peaks_out     <span class="op">=</span> <span class="st">"mm39_peaks.rds"</span>,</span>
<span>                    map_out       <span class="op">=</span> <span class="st">"mm39_map.rds"</span>, </span>
<span>                    genoprobs     <span class="op">=</span> <span class="va">probs</span>, </span>
<span>                    samp_meta     <span class="op">=</span> <span class="va">demo_meta</span>,                   <span class="co"># Make sure that your ID column is named "ID"</span></span>
<span>                    expr_mats     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"mESC"</span> <span class="op">=</span> <span class="va">demo_counts</span><span class="op">)</span>,  <span class="co"># Pass your counts as a named list</span></span>
<span>                    covar_factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span><span class="op">)</span>,</span>
<span>                    gridFile      <span class="op">=</span> <span class="va">gridfile</span>, </span>
<span>                    annots        <span class="op">=</span> <span class="va">demo_annot</span>,                  <span class="co"># See the Annotations section above for notes on this</span></span>
<span>                    save          <span class="op">=</span> <span class="st">"ro"</span><span class="op">)</span>                        <span class="co"># return only - other options are "so" (save only) and "sr" (save and return)</span></span>
<span></span>
<span><span class="co">## Passing file locations</span></span>
<span><span class="va">map_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapQTL.html">mapQTL</a></span><span class="op">(</span>outdir        <span class="op">=</span> <span class="st">"../../vignette/"</span>, </span>
<span>                    peaks_out     <span class="op">=</span> <span class="st">"mm39_peaks.rds"</span>,</span>
<span>                    map_out       <span class="op">=</span> <span class="st">"mm39_map.rds"</span>, </span>
<span>                    genoprobs     <span class="op">=</span> <span class="st">"path/to/saved/probs.rds"</span>,</span>
<span>                    samp_meta     <span class="op">=</span> <span class="st">"path/to/sample/metadata"</span>,</span>
<span>                    expr_mats     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"path/to/counts1"</span>,<span class="st">"path/to/coounts/2"</span><span class="op">)</span>, </span>
<span>                    covar_factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span><span class="op">)</span>,</span>
<span>                    gridFile      <span class="op">=</span> <span class="st">"path/to/custom/grid"</span>,</span>
<span>                    annots        <span class="op">=</span> <span class="st">"path/to/annotations"</span>,</span>
<span>                    save          <span class="op">=</span> <span class="st">"ro"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="mediation">Mediation<a class="anchor" aria-label="anchor" href="#mediation"></a>
</h3>
<p>For mediation we use the peaks, mapping information (which includes
the rankZ normalized counts), and annotations to identify targets which
in turn allows us to identify phenotypes that act as mediators for peaks
above a suggestive LOD. This is done using
<code><a href="https://rdrr.io/pkg/intermediate/man/mediation.scan.html" class="external-link">intermediate::mediation.scan</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Unevaluated Code Chunk</span></span>
<span><span class="va">med_res</span> <span class="op">&lt;-</span> <span class="fu">run_mediate</span><span class="op">(</span>peaks   <span class="op">=</span> <span class="va">map_peaks</span><span class="op">$</span><span class="va">peaks_list</span>, </span>
<span>                       mapping <span class="op">=</span> <span class="va">map_peaks</span><span class="op">$</span><span class="va">maps_list</span>,</span>
<span>                       outdir  <span class="op">=</span> <span class="st">"../../vignette"</span>, </span>
<span>                       annots  <span class="op">=</span> <span class="va">demo_annot</span>,</span>
<span>                       suggLOD <span class="op">=</span> <span class="fl">7</span>, </span>
<span>                       med_out <span class="op">=</span> <span class="st">"mm39_mediation.rds"</span>, </span>
<span>                       save    <span class="op">=</span> <span class="st">"ro"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="effects">Effects<a class="anchor" aria-label="anchor" href="#effects"></a>
</h3>
<p>To identify the founder effects at peaks of interest, we use the
peaks, and mapping information along with a suggestive LOD threshold to
ensure that the reported effects are for relevant peaks only. This is
done using <code><a href="https://rdrr.io/pkg/qtl2/man/scan1blup.html" class="external-link">qtl2::scan1blup</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Unevaluated Code Chunk</span></span>
<span><span class="va">effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qtl_effects.html">qtl_effects</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="va">map_peaks</span><span class="op">$</span><span class="va">maps_list</span>, </span>
<span>                       peaks   <span class="op">=</span> <span class="va">map_peaks</span><span class="op">$</span><span class="va">peaks_list</span>,</span>
<span>                       suggLOD <span class="op">=</span> <span class="fl">7</span>, </span>
<span>                       outdir  <span class="op">=</span> <span class="st">"../../vignette"</span>, </span>
<span>                       outfile <span class="op">=</span> <span class="st">"mm39_effects.rds"</span>, </span>
<span>                       save    <span class="op">=</span> <span class="st">"ro"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="submitting-to-a-job-scheduler">Submitting to a Job Scheduler<a class="anchor" aria-label="anchor" href="#submitting-to-a-job-scheduler"></a>
</h2>
<p>Currently, <code>QTLretrievR</code> is set up to identify available
cores in a linux, mac, or windows environment. It is also able to
identify the number of cores if you are submitting a job through a SLURM
scheduler. Just make sure to include both of the following lines in your
<code>SBATCH</code>:</p>
<pre><code><span><span class="co">#SBATCH --ntasks=N</span></span>
<span><span class="co">#SBATCH --cpus-per-task=X</span></span></code></pre>
<p>When the number of cores is calculated for parallelization purposes,
the total number of cores that are identified are <code>N*X</code>. If
you don’t include both, and are submitting with a SLURM scheduler, it
may not identify the correct number of cores.</p>
<p>If you are using any other scheduler, make sure that you specify the
total number of cpus for the whole job. If applicable the number of
processes should be 1.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah B. Dewey, Selcan Aydin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
