<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate mapping data and peaks for QTL analysis — mapQTL • QTLretrievR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Generate mapping data and peaks for QTL analysis — mapQTL"><meta name="description" content="This function performs QTL mapping across multiple tissues, generating
genome-wide LOD scores and identifying significant peaks. It supports
parallel processing and flexible input formats for expression data and
covariates."><meta property="og:description" content="This function performs QTL mapping across multiple tissues, generating
genome-wide LOD scores and identifying significant peaks. It supports
parallel processing and flexible input formats for expression data and
covariates."><meta property="og:image" content="https://deweyhannah.github.io/QTLretrievR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">QTLretrievR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/QTLretrievR.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/DOAttie500_LiverLipids.html">Liver Lipids Example</a></li>
    <li><a class="dropdown-item" href="../articles/DOAttie500_Islet.html">Islet Transcriptomics Example</a></li>
  </ul></li>
<li class="nav-item"><h6 class="dropdown-header" data-toc-skip>Reference</h6></li>
      </ul><ul class="navbar-nav"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Generate mapping data and peaks for QTL analysis</h1>

      <div class="d-none name"><code>mapQTL.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs QTL mapping across multiple tissues, generating
genome-wide LOD scores and identifying significant peaks. It supports
parallel processing and flexible input formats for expression data and
covariates.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mapQTL</span><span class="op">(</span></span>
<span>  <span class="va">genoprobs</span>,</span>
<span>  <span class="va">samp_meta</span>,</span>
<span>  <span class="va">expr_mats</span>,</span>
<span>  <span class="va">covar_factors</span>,</span>
<span>  thrA <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  thrX <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  gridFile <span class="op">=</span> <span class="va">gridfile</span>,</span>
<span>  localRange <span class="op">=</span> <span class="fl">2e+06</span>,</span>
<span>  outdir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peaks_out <span class="op">=</span> <span class="st">"peaks.rds"</span>,</span>
<span>  map_out <span class="op">=</span> <span class="st">"map.rds"</span>,</span>
<span>  annots <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  total_cores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  save <span class="op">=</span> <span class="st">"sr"</span>,</span>
<span>  rz <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  phys <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genoprobs">genoprobs<a class="anchor" aria-label="anchor" href="#arg-genoprobs"></a></dt>
<dd><p>Genotype probabilities object (qtl2-formatted), or path to
.rds file containing one.</p></dd>


<dt id="arg-samp-meta">samp_meta<a class="anchor" aria-label="anchor" href="#arg-samp-meta"></a></dt>
<dd><p>Sample metadata. Either a string pointing to the file,
or the object itself.</p></dd>


<dt id="arg-expr-mats">expr_mats<a class="anchor" aria-label="anchor" href="#arg-expr-mats"></a></dt>
<dd><p>List of normalized count matrices (objects), or character
paths to the file. One matrix per tissue. The order <em>must match</em>
the tissue order in <code>genoprobs</code>. <strong>NOTE</strong> If using <code>rz = TRUE</code> then
phenotypes should be the columns in the expression matrices.</p></dd>


<dt id="arg-covar-factors">covar_factors<a class="anchor" aria-label="anchor" href="#arg-covar-factors"></a></dt>
<dd><p>Additive covariate factors. These need to be columns
in the factor metadata.</p></dd>


<dt id="arg-thra">thrA<a class="anchor" aria-label="anchor" href="#arg-thra"></a></dt>
<dd><p>Minimum reported LOD threshold for autosomes. Default is 5.</p></dd>


<dt id="arg-thrx">thrX<a class="anchor" aria-label="anchor" href="#arg-thrx"></a></dt>
<dd><p>Minimum reported LOD threshold for X chromosome. Default is 5.</p></dd>


<dt id="arg-gridfile">gridFile<a class="anchor" aria-label="anchor" href="#arg-gridfile"></a></dt>
<dd><p>Genome Grid. Path to location or object. Defaults to
75k grid loaded with package.</p></dd>


<dt id="arg-localrange">localRange<a class="anchor" aria-label="anchor" href="#arg-localrange"></a></dt>
<dd><p>Definition of "local" in bp. Default is 2e6.</p></dd>


<dt id="arg-outdir">outdir<a class="anchor" aria-label="anchor" href="#arg-outdir"></a></dt>
<dd><p>Directory to save output files. Default is <code>NULL</code>.</p></dd>


<dt id="arg-peaks-out">peaks_out<a class="anchor" aria-label="anchor" href="#arg-peaks-out"></a></dt>
<dd><p>String indicating the name for output peaks file. This file
will be saved in <code>.rds</code> format and will be used as an input for downstream
analysis. Should end  in <code>.rds</code>. Default is "<code>gxe_peaks.rds</code>"</p></dd>


<dt id="arg-map-out">map_out<a class="anchor" aria-label="anchor" href="#arg-map-out"></a></dt>
<dd><p>String indicating the name of the output file containing QTL
mapping results. This file will be saved in <code>.rds</code> format and will be used
in downstream analyses. Should end in <code>.rds</code>. Default is "<code>map.rds</code>"</p></dd>


<dt id="arg-annots">annots<a class="anchor" aria-label="anchor" href="#arg-annots"></a></dt>
<dd><p>Annotations file. Contains mapping information for phenotypes.
Dataframe, or tsv. Columns must include "id", "symbol", "chr", "start", "end".</p></dd>


<dt id="arg-total-cores">total_cores<a class="anchor" aria-label="anchor" href="#arg-total-cores"></a></dt>
<dd><p>Number of available cores to use for parallelization.
Default is <code>NULL</code>.</p></dd>


<dt id="arg-save">save<a class="anchor" aria-label="anchor" href="#arg-save"></a></dt>
<dd><p>Indicates object return/save behavior. One of
<code>c("sr", "so", "ro")</code>; save &amp; return, save only, return only.
Default is "sr".</p></dd>


<dt id="arg-rz">rz<a class="anchor" aria-label="anchor" href="#arg-rz"></a></dt>
<dd><p>Logical. Set to <code>TRUE</code> if expression data is already
rankZ-transformed. Default is <code>FALSE</code>.</p></dd>


<dt id="arg-phys">phys<a class="anchor" aria-label="anchor" href="#arg-phys"></a></dt>
<dd><p>Logical. if <code>TRUE</code>, use the physical map for peak calling;
otherwise use the genomic map. Default is <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p><ul><li><p>maps_listA list of dataframes and lists to that can be used for
future analyses and in other functions </p><ul><li><p>qtlprobsGenome probabilities in qtl2format</p></li>
<li><p>covar_listlist of covariate matrices for each tissue</p></li>
<li><p>expr_listOriginal normalized expression data for each tissue</p></li>
<li><p>exprZ_listRank Z-transformed expression data for each tissue</p></li>
<li><p>kinship_locoKinship matrix calculated using the "loco"
option in <code><a href="https://rdrr.io/pkg/qtl2/man/calc_kinship.html" class="external-link">qtl2::calc_kinship</a></code></p></li>
<li><p>gmapGenomic map of markers</p></li>
<li><p>map_dat2Combined genomic and physical map of markers</p></li>
<li><p>pmapPhysical map of markers</p></li>
<li><p>tissue_sampMetadata broken down for each tissue</p></li>
</ul></li>
<li><p>peaks_listA list of peaks list for each tissue.</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah B. Dewey, Selcan Aydin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.9000.</p>
</div>

    </footer></div>





  </body></html>

